tests/unit/parallel/algorithms/is_heap_until_tests.hpp

```diff
@@ -0,0 +1,281 @@
+    }
+    catch(...) {
+        HPX_TEST(false);
+    }
+
+    HPX_TEST(caught_exception);
+}
+
+template <typename ExPolicy, typename IteratorTag>
+void test_is_heap_until_exception_async(ExPolicy policy, IteratorTag)
+{
+    typedef std::vector<int>::iterator base_iterator;
+    typedef test::test_iterator<base_iterator, IteratorTag> iterator;
+
+    std::vector<int> c(10007);
+    std::iota(boost::begin(c), boost::end(c), std::rand());
+    std::make_heap(boost::begin(c), boost::end(c));
+
+    bool caught_exception = false;
+    bool returned_from_algorithm = false;
+    try {
+        auto f =
+            hpx::parallel::is_heap_until(policy,
+                iterator(boost::begin(c)), iterator(boost::end(c)),
+                throw_always());
+        returned_from_algorithm = true;
+        f.get();
+
+        HPX_TEST(false);
+    }
+    catch(hpx::exception_list const& e) {
+        caught_exception = true;
+        test::test_num_exceptions<ExPolicy, IteratorTag>::call(policy, e);
+    }
+    catch(...) {
+        HPX_TEST(false);
+    }
+
+    HPX_TEST(caught_exception);
+    HPX_TEST(returned_from_algorithm);
+}
+
+///////////////////////////////////////////////////////////////////////////////
+template <typename ExPolicy, typename IteratorTag>
+void test_is_heap_until_bad_alloc(ExPolicy policy, IteratorTag)
+{
+    static_assert(
+        hpx::parallel::execution::is_execution_policy<ExPolicy>::value,
+        "hpx::parallel::execution::is_execution_policy<ExPolicy>::value");
+
+    typedef std::vector<int>::iterator base_iterator;
+    typedef test::test_iterator<base_iterator, IteratorTag> iterator;
+
+    std::vector<int> c(10007);
+    std::iota(boost::begin(c), boost::end(c), std::rand());
+    std::make_heap(boost::begin(c), boost::end(c));
+
+    bool caught_bad_alloc = false;
+    try {
+        hpx::parallel::is_heap_until(policy,
+            iterator(boost::begin(c)), iterator(boost::end(c)),
+            throw_bad_alloc());
+
+        HPX_TEST(false);
+    }
+    catch(std::bad_alloc const&) {
+        caught_bad_alloc = true;
+    }
+    catch(...) {
+        HPX_TEST(false);
+    }
+
+    HPX_TEST(caught_bad_alloc);
+}
+
+template <typename ExPolicy, typename IteratorTag>
+void test_is_heap_until_bad_alloc_async(ExPolicy policy, IteratorTag)
+{
+    typedef std::vector<int>::iterator base_iterator;
+    typedef test::test_iterator<base_iterator, IteratorTag> iterator;
+
+    std::vector<int> c(10007);
+    std::iota(boost::begin(c), boost::end(c), std::rand());
+    std::make_heap(boost::begin(c), boost::end(c));
+
+    bool caught_bad_alloc = false;
+    bool returned_from_algorithm = false;
+    try {
+        auto f =
+            hpx::parallel::is_heap_until(policy,
+                iterator(boost::begin(c)), iterator(boost::end(c)),
+                throw_bad_alloc());
+        returned_from_algorithm = true;
+        f.get();
+
+        HPX_TEST(false);
```

Comment:
It's far out of scope for this PR, but it would be nice if there would be something like a `HPX_TEST_FAIL();` macro, like GTest provides one.
Just checked it and something like this doesn't seem to exist.